PlayFabSDK/source/Qos/RegionPinger.cs

```diff
@@ -3,76 +3,136 @@ namespace PlayFab.QoS
 {
     using System;
     using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Threading;
     using System.Threading.Tasks;
-    using MultiplayerModels;
-#if NET45 || NETCOREAPP2_0
+#if (NETSTANDARD && !NETSTANDARD1_1) || NETFRAMEWORK || NETCOREAPP
     using System.Net;
     using System.Net.Sockets;
-#endif
 
-    public class RegionPinger
+    public class RegionPinger : IRegionPinger
     {
-        private const int PingCount = 5;
         private const int PortNumber = 3075;
         private const int UnknownLatencyValue = int.MaxValue;
-        private readonly byte[] _initialHeader = { 0xFF, 0xFF };
-        private readonly byte[] _subsequentHeader = { 0x00, 0x00 };
+        private static readonly byte[] _initialHeader = {0xFF, 0xFF};
+        private static readonly byte[] _subsequentHeader = {0x00, 0x00};
+
+        private readonly int _numTimeoutsForError;
         private readonly string _hostNameOrAddress;
         private readonly string _region;
+        private readonly int _timeoutMs;
 
-        public RegionPinger(string hostNameOrAddress, string region)
+        private int _numTimeouts;
+        private readonly List<int> latencyMeasures;
+
+        public RegionPinger(string hostNameOrAddress, string region, int timeoutMs, int numTimeoutsForError,
+            int expectedPingsPerRegion = 0)
```

Comment:
which one needs units? expectedPingsPerRegion the unit is "pings". The default of 0 is provided because it is not strictly required it is just a performance improvement to presize the list.

---
In reply to: [376144716](https://github.com/PlayFab/CSharpSDK/pull/111#discussion_r376144716) [](ancestors = 376144716)